pkg/workload/kv/dynamic.go

```diff
@@ -0,0 +1,550 @@
+	return func(s []int64) []int64 {
+		nSlots := len(s)
+		newSlots := make([]int64, nSlots)
+
+		// Sanity check that we don't have a negative or zero chunk size.
+		if chunk < 1 {
+			chunk = 1
+		}
+
+		nChunks := int(math.Ceil(float64(nSlots) / float64(chunk)))
+		permArray := rand.Perm(nChunks)
+
+		for i := 0; i < nChunks; i++ {
+			for j := 0; j < chunk && i*chunk+j < nSlots; j++ {
+				from := permArray[i]*chunk + j
+				to := i*chunk + j
+				newSlots[to] = s[from]
+			}
+		}
+		return newSlots
+	}
+}
+
+// identityFn returns a paritalMapFn that returns the same array that it is
+// given..
+func identityFn() partialMapFn {
+	return func(s []int64) []int64 {
+		return s
+	}
+}
+
+// slots provides methods to apply partial and uniform mapping to keys. Keys
+// are held in slots, which may be rearranged to change the resulting location
+// of a key. Initially, slot 0 holds key 0, slot 1 holds key 1... slot n holds
+// key n.
+type slots interface {
+	// uniformMap applies a function to all slots.
+	uniformMap(uniformMapFn)
+	// partialMap applies a function to a subset of underlying slots.
+	partialMap(partialMapFn)
+	// key returns the key held at the slot passed in.
+	key(slot int64) int64
+	// enforceUniform removes any individual slot mapping. All future calls to
+	// partialMap will have no effect.
+	enforceUniform()
+}
+
+// slotMap is an implementation of the slots interface.
+type slotMap struct {
+	slots        []int64
+	uniformDelta int64
+	nSlots       int64
+	noSlot       bool
+}
+
+func (sm *slotMap) uniformMap(f uniformMapFn) {
+	sm.uniformDelta = f(sm.uniformDelta)
+}
+
+func (sm *slotMap) partialMap(f partialMapFn) {
+	// Disallow partial mapping when no slot has been declared.
+	if sm.noSlot {
+		return
+	}
+
+	// Lazily initialize the slot array when a partial map is applied for the
+	// first time.
+	if len(sm.slots) < int(sm.nSlots) {
+		sm.slots = make([]int64, sm.nSlots)
+		for i := range sm.slots {
+			sm.slots[i] = int64(i)
+		}
+	}
+
+	sm.slots = f(sm.slots)
+}
+
+func (sm *slotMap) key(slot int64) int64 {
+	var slotKey int64
+	slot = slot % sm.nSlots
+	if sm.noSlot {
+		slotKey = slot
+	} else {
+		slotKey = sm.slots[slot]
+	}
+
+	return (slotKey + sm.uniformDelta) % sm.nSlots
+}
+
+func (sm *slotMap) enforceUniform() { sm.noSlot = true }
+
+type distribution interface {
+	key(width int64) int64
+}
+
+func newDistribution(config *kv, random *rand.Rand) distribution {
+	if config.sequential {
+		return &sequentialD{}
+	}
+	if config.zipfian {
+		return &zipfD{rand: random}
+	}
+	return &uniformD{rand: random}
+}
+
+type zipfD struct {
+	rand *rand.Rand
+}
+
+func (z *zipfD) key(width int64) int64 {
+	zipf := newZipf(1.1, 1, uint64(width))
+	return width - int64(zipf.Uint64(z.rand))
+}
+
+type uniformD struct {
+	rand *rand.Rand
+}
+
+func (u *uniformD) key(width int64) int64 {
+	return u.rand.Int63n(width)
+}
+
+type sequentialD struct {
+	seq int64
+}
+
+func (s *sequentialD) key(width int64) int64 {
+	s.seq++
```

Comment:
Do these key generators need to be thread-safe, either through the use of atomic operations or through the use of a mutex?

Reply:
Actually, it doesn't look like we're even sharing these dynamic generators across concurrent threads. Instead, each worker uses its own generator. For some of the random distributions, this may be ok, but for a stateful generator like "moving hotkey", I would think we'd want the generator to be shared so that a single hotspot with some internal concurrency moves across the keyspace.

Reply:
I don't think they're shared in any kv generator.

In this case, the mapping is deterministic and symmetric (except when it's triggered by #keys generated, where there will be a little lag between).

So I suppose the issue is the underlying static distribution? Since it's not shared it will have key amplification (same P(key), just everyone picks it, as opposed to spreading)?


I am somewhat banking on them being not threadsafe but if need be I can put a mutex on remapping, it just doesn't seem necessary unless this amplification is really an issue.

I'll look into it further later this week.

Reply:
So I think I misunderstood - you mean sharing a sequence ref, which is atomically incremented? That seems reasonable - I'll do some updates.